### PR TITLE
Enhancements for database pool

### PR DESCRIPTION
**Pull request by WeaxMe**
1. Added possibility for check if database pool is closed.
2. Created database pool factory which allows manage database pools associated with users

### CODE REVIEW

**Comment by tglman** [comment]
Hi @WeaxMe,

for the 1. definitely ok for me.

for the 2. well it is a bit more complex:

As today OrientDB has two main entry points, the `OrientDB` class and the `ODatabasePool` class with the second one actually using internally the first one (or just the underlying implementation `OrientDBInternal` ), both of the class are actually designed to be a twin layer that hide the implementations detail provided by `OrientDBInternal` so to respect the current design the logic that you are providing with the `ODatabasePoolFactory` should actually be implemented by `OrienDBIternal` and it's implementions, `OrientDBEmbedded`, `OrientDBRemote` and `OrientDBDistributed`.

So as replacement of `ODatabasePoolFactory` I would see a new API on `OrientDB` class with the signature like:
```
ODatabasePool cachedPool(String database,String user, String password)
```
@wolf4ood ,  @luigidellaquila, @laa,  What do you think ?

P.S.

Your implementation of `ODatabasePoolFactory` is definitely good for a project that use OrientDB, I'm just trying to keep the number of API low and coherent.

**Comment by WeaxMe** [comment]
Hi @tglman 
Thank you for feedback.
OK I understand what need to do. If you are OK I will implement this things which you described

**Comment by tglman** [comment]
Hi @WeaxMe,

Yes that would be perfect, be aware that you may need to duplicate a bit the logic between `OrientDBEmbedded` and `OrientDBRemote`, and that will be OK.

Another small detail, if is possible can you try to avoid to store the password in the `PoolIdentity`, database and user should be enough as identifiers, we do keep the password stored in memory in other places, but we are trying to reduce this and not store the password anywhere in future.

Regards

**Comment by tglman** [comment]
Hi @WeaxMe,

This looks really good now, could you add the same implementation in OrientDBRemote?  can just be implemented in the same way is implemented in embedded using the same factory implementation, if you add as well the same tests that you added to OrientDBRemoteTest should be complete.

I see that you changed the pom.xml  as well, is there any reason for that change ? the diff shows a lot of changes so I cannot really say if is a wanted change or a formatting change, if is a formatting change can you just remove it ?

Thank you very much for your work !!

**Comment by WeaxMe** [comment]
Hi @tglman yeh I will implement all what I say. I just didn't have a time for complete it. I think that during this weekends I will finish work.

Sure I will check pom.xml

**Comment by tglman** [comment]
Hi @WeaxMe,

Any update ?
Regards

**Comment by PhantomYdn** [comment]
@tglman , @WeaxMe is working on that. We at Orienteer put on-hold for a while upgrade to OrientDB 3.x, but now we are back:) Will update you soon.

**Comment by WeaxMe** [comment]
Hello @tglman 
I added support for OrientDBRemote. 
Please see changes

**Comment by WeaxMe** [comment]
Hello @tglman 
Sorry about wait for my changes.

What do you think about new changes in PR?

**Comment by PhantomYdn** [comment]
@tglman , @laa , could you please help with the next steps? Really looking for having this functionality merged.

**Comment by tglman** [comment]
Hi All,

Sorry for the delay, i will double check and merge it.

Regards

File: core/src/main/java/com/orientechnologies/orient/core/db/OCachedDatabasePoolFactoryImpl.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    checkForClose();
+
+    String key = OSecurityManager.instance().createSHA256(database + username + password);
+    ODatabasePoolInternal pool = getByKey(key);
+
+    if (pool != null) {
+      return pool;
+    }
+
+    return poolCache.computeIfAbsent(key, hash -> {
```
[the remark this answers is not in the document]
Comment (WeaxMe) [reply core/src/main/java/com/orientechnologies/orient/core/db/OCachedDatabasePoolFactoryImpl.java]: Done. I added check for closed pool

File: core/src/main/java/com/orientechnologies/orient/core/db/OCachedDatabasePoolFactoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+      OrientDBConfig config = OrientDBConfig.builder()
+              .addConfig(OGlobalConfiguration.DB_POOL_MAX, maxPoolSize)
+              .build();
+
+      if (parentConfig != null) {
+        config.setParent(parentConfig);
+      }
+
+      return new ODatabasePoolImpl(orientDB, database, username, password, config);
```
[the remark this answers is not in the document]
Comment (WeaxMe) [reply core/src/main/java/com/orientechnologies/orient/core/db/OCachedDatabasePoolFactoryImpl.java]: Yeah. This make sense. Let's create issue and I will complete it in next time

**Comment by WeaxMe** [comment]
I added requested changes. Please see code

**Comment by PhantomYdn** [comment]
@tglman , could you please review this PR after recent changes?

**Pull request merged**